Add symbol to tell the linker that new libsystemd version is needed

keszybz opened the pull request.

dvdhrm commented:
I like it!

I would change the `vtable_format_reference` member to be a pointer to an integer, though, and then make `sd_bus_object_vtable_format_243` an `extern const int` as well. It feels more natural than having a no-op function, doesn't it?

I also very much like that you now correctly iterate the vtable based on the element-size.

Other comments inline.

Thanks a lot!

dvdhrm commented on src/libsystemd/sd-bus/bus-objects.c:
```diff
@@ -1701,7 +1701,7 @@ struct sd_bus_vtable_original {
+        if (vtable[0].x.start.element_size < VTABLE_ELEMENT_SIZE ||
             !vtable[0].x.start.vtable_format_reference)
                 return 0;
         return vtable[0].x.start.features;
```
I don't understand this change. Imagine a future modification to `struct sd_bus1_vtable`, which increases its size and as such `VTABLE_ELEMENT_SIZE` will increase as well. If you now run an old binary against this newly compiled library, `vtable_features()` will return 0, even though the old binary might have actually put features into the structure.

Or am I reading this wrong? Is the intention to not update `VTABLE_ELEMENT_SIZE` here when it changes again?

keszybz replied:
Yes, the idea is to update the check to use the current (as of now) size in the future. I renamed the macros and added a comment to make this clear.

keszybz commented:
Thanks for the review.

Updated. I changed the new symbol to be an `extern unsigned const`. This is indeed nicer. There were some cascading small changes required because our tests assign all symbols to a `void *` and such. I also renamed some macros and added comments to make things clearer. PTAL.

poettering commented on src/libsystemd/sd-bus/bus-objects.c:
```diff
@@ -1745,8 +1742,8 @@ static int add_object_vtable_internal(
-                      vtable[0].x.start.element_size == VTABLE_ELEMENT_SIZE,
+        assert_return(vtable[0].x.start.element_size == VTABLE_ELEMENT_SIZE_221 ||
+                      vtable[0].x.start.element_size >= VTABLE_ELEMENT_SIZE_242,
                       -EINVAL);
```
hmm, but now we'd silently accept vtables that might use future features we don't support which might create problems.

I figure there should be a vtable_features() check here, that validates that there are no bits set in it we don't understand

keszybz replied:
No, that would be useless. If the table is extended in a backwards-compatible way, the old code will work fine with the new tables. If the table is extended in a non-backwards compatible way, we'll need to add a new symbol and/or bump symbol versions.

Consider the last addition: it didn't work in this way because of the overzealous size check in the old code. But apart from that, it should have been fully backwards-compatible. If we were later to e.g. add another field in the structure that would provide some additional information (but the table would still be usable without it), the code in this PR should work fine with it.

poettering commented on man/sd_bus_add_object_vtable.xml:
```diff
@@ -0,0 +1,467 @@
+  <refsect1>
+    <title>Description</title>
+
+    <para><function>sd_bus_add_object_vtable()</function> is used to declare attributes for path
```
path → bus object path

poettering commented on man/sd_bus_add_object_vtable.xml:
```diff
@@ -0,0 +1,467 @@
+    <title>Description</title>
+
+    <para><function>sd_bus_add_object_vtable()</function> is used to declare attributes for path
+    <parameter>path</parameter> connected to the bus object <parameter>bus</parameter> under the interface
```
hmm using "bus object" is a bit confusing here, since bus objects could either be the bus connection object (which is what you mean here) or the any of the dbus objects exposed on a dbus service... hence: "bus object" → "bus connection", maybe?

poettering commented on man/sd_bus_add_object_vtable.xml:
```diff
@@ -0,0 +1,467 @@
+    <constant>SD_BUS_METHOD_WITH_NAMES_OFFSET()</constant>, and signal declarations using
+    <constant>SD_BUS_SIGNAL_WITH_NAMES()</constant> or <constant>SD_BUS_SIGNAL()</constant>, see below. The
+    <replaceable>userdata</replaceable> parameter contains a pointer that will be passed to various
+    callback. It may be specified as <constant>NULL</constant> if no value is necessary.</para>
```
callback → callback functions?

poettering commented on man/sd_bus_add_object_vtable.xml:
```diff
@@ -0,0 +1,467 @@
+    <refsect2>
+      <title>The <structname>sd_bus_vtable</structname> entries</title>
+
+      <para>The following elements are allowed:</para>
```
maybe say here that the array consists of structures, but that these structures should never be filled in manually, but always through these macros. Also, say that these macros are in fact macros

poettering commented on man/sd_bus_add_object_vtable.xml:
```diff
@@ -0,0 +1,467 @@
+    entries).</para>
+
+    <para>For both functions, a match slot is created internally. It will be returned in the
+    <replaceable>slot</replaceable> output parameter if that is non-null.</para>
```
maybe say that sd_bus_slot_unref() should be used to drop the vtables again

keszybz commented:
I updated the man page with the suggested changes. I also noticed an error in the description of SD_BUS_VTABLE_START: it does take some flags.
